lamagraph-core/src/Core/Reducer.hs

```diff
@@ -0,0 +1,199 @@
+  -- | Left `Node` to handle.
+  LoadedNode numberOfPorts ->
+  -- | Right `Node` to handle.
+  Maybe (LoadedNode numberOfPorts) ->
+  -- | Next possible Node and Address to handle and result data.
+  (Maybe (LoadedNode numberOfPorts, Address), DataToStore maxNumOfNodesToStore maxNumOfEdgesToStore numberOfPorts)
+handle leftNode rightNode =
+  selectNextLeftNode $ DataToStore (markAllInnerEdges nodes) emptyEdges
+ where
+  emptyNodes = repeat Nothing :: Vec maxNumOfNodesToStore _
+  emptyEdges = repeat Nothing :: Vec maxNumOfEdgesToStore _
+  nodes = case rightNode of
+    Nothing -> Just leftNode +>> emptyNodes
+    Just _ ->
+      -- if isActive leftNode n
+      rightNode +>> Just leftNode +>> emptyNodes -- reduction rules applied here
+
+-- | Type to indicate that all ports that could be handled locally are over.
+data NonVisitedOver = No | Yes
+
+-- | Transition function in mealy automaton.
+mealyFunction ::
+  ( KnownDomain dom
+  , HiddenClockResetEnable dom
+  , KnownNat maxNumOfNodesToStore
+  , KnownNat maxNumOfEdgesToStore
+  , KnownNat numberOfPorts
+  ) =>
+  -- | State of reducer and automaton.
+  ReducerState ->
+  -- | Input.
+  ( -- Left and right `Node` to reduce.
+    (Maybe (LoadedNode numberOfPorts), Maybe (LoadedNode numberOfPorts))
+  , ReducerStatus
+  , Maybe Address
+  -- Address from memory manager if reducer cannot find next `Address` locally.
+  ) ->
+  -- | (state, output).
+  ( ReducerState
+  , ( Maybe (DataToStore maxNumOfNodesToStore maxNumOfEdgesToStore numberOfPorts)
+    , ReducerStatus
+    , Maybe (LoadedNode numberOfPorts)
+    , -- Next left `Node`
+      Maybe Address
+    , -- Next `Address` to load
+      NonVisitedOver
+    )
+  )
+mealyFunction
+  state
+  ( (mbLeftNode, mbLoadedNode)
+    , status
+    , addressOfNodeWithUnvisitedEdges
+    ) =
+    case state of
+      Initial addressOfFirstNodeToLoad -> (Empty, (Nothing, Work, Nothing, Just addressOfFirstNodeToLoad, No))
+      Empty ->
+        case mbLoadedNode of
+          Nothing ->
+            (Failed, (Nothing, ErrorEmptyLeftNode, Nothing, Nothing, No))
+          Just loadedNode ->
+            case addressToLoad of
+              Just _ ->
+                (OneNodeToLoad, (Nothing, Work, Just loadedNode, addressToLoad, No))
+              Nothing ->
+                case mbLeftNode of
+                  Just leftNode' ->
+                    case infoAboutNextNodes of
+                      Just (_, _) ->
+                        (Failed, (Nothing, ErrorHandleSingleNode, Nothing, Nothing, No))
+                      Nothing ->
+                        (Empty, (Just dataToStore, Work, Nothing, addressOfNodeWithUnvisitedEdges, Yes))
+                   where
+                    (infoAboutNextNodes, dataToStore) = handle leftNode' Nothing
+                  Nothing ->
+                    (Failed, (Nothing, ErrorEmptyLeftNode, Nothing, Nothing, No))
+           where
+            addressToLoad = selectAddressToLoad loadedNode
```

Comment:
I think this logic must have its own function, maybe functionS

Reply:
Why? I don't think this is a significant simplification

Reply:
IMO this code is hard to read because of gigantic nesting of case expressions and splitting it might help with readability.